docs/06-contributors/999-rfcs/000-stories/story-13.md

```diff
@@ -0,0 +1,169 @@
+# User Story 13 - Task List with uuid, Json and duration
+
+> **Status**: Expected released on 2023/02/16
+
+The following code is an iteration of Wing's task list.
+This version of it includes the following functional changes:
+- Using npm's uuid package inflight instead of the counter resource 
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions 
+  - +100 if we are able to do the same for azure as well
+- Least privileged permissions are granted
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete 
+- Console should be able to show all resources, and interact with them
+
+
+
+```ts (wing)
+bring cloud;
+// need to run: npm install uuid --save
+bring uuid;
+
+/**
+ * Represents a cloud task list.
+ */
+resource TaskList {
+  /** stores the tasks */
+  _bucket: cloud.Bucket;
+  
+  init() {
+    this._bucket = new cloud.Bucket();
+  }
+
+  /** 
+   * Adds a task to the task list.
+   * @returns The ID of the new task.
+   */
+  inflight add_task(title: str): str {
+    let id = uuid.v4();
+    let j = Json { 
+      title: title, 
+    };
+    print("adding task ${id} with data: ${j}"); //j should be printed out nicely 
+    this._bucket.put(id, j.to_str());
+    return id;
+  }
+
+  /** 
+   * Gets a task from the task list.
+   * @param id - the id of the task to return
+   * @returns the title of the task (optimistic)
+   */
+  inflight get_task(id: str): Json {
+    let blob = this._bucket.get(id);
```

Comment:
```suggestion
    let blob = this._bucket.get_json(id);
```

Reply:
I don't expect `str` to be a subtype of `Json` 

```ts (wing)
print_json(j: Json) {
   if (j.some_bool_value) {
      print(j.some_str_value)
   }
}

print_json("I am not a json");
```

Reply:
Ok, maybe I jumped the gun here -- so even though `x` in my example above could be a string-ish value, maybe it's not correct to say `Json "hi"` is the same as `"hi"`.

Eyal raised this question ("if string is a subtype of json, then is this allowed?") that stumped me...

```
"hello".some_bool_value
```

Reply:
I don't think `bucket.put()` should always take a JSON because there are definitely use cases to store objects that are not JSON (e.g. binary blobs).

Maybe in the future we should define a "Serializable" interface ("trait") and have both `str` and `Json` (and `Blob`) implement it, and then `put()` just takes a `Serializable`.

Reply:
@eladb I totally agree - I don't think only accepting `Json` should be our endgame. My original comment was just trying to point out that if our choices for story 13 are between the following:

```
// API A
resource Bucket {
  ...
  inflight put(key: str, value: str) { ... }
  inflight put_json(key: str, value: Json) { ... }
}
```

```
// API B
resource Bucket {
  ...
  inflight put(key: str, value: Json) { ... }
}
```

Then API B might be cleaner and less confusing for the beta / pre-beta.

Reply:
But then how would you store a string in the bucket?

`"hello".to_json()` returns `"hello"` (with quotes).

How about we do API A and rename `put()` to `put_str()`.

(basically `put_json()` and `put_str()` and `get_json()` and `get_str()`).

Reply:
Ahh, gotcha - `"hello"` can be parsed back into JSON while `hello` cannot. Not sure why I got so tripped up on this... yeah separate methods for these make sense to me now.